### PR TITLE
Put all plugins behind an "API".

### PR DESCRIPTION
**Pull request by ghost**
This is a pure, undiluted reality check (all the stuff our own plugins currently rely on).

From here, it is possible to make some random changes and walk in circles, but I was wondering whether can agree on a direction (what is it that we want; what is it what we reject).

### CODE REVIEW

**Comment by ejgallego** [comment]
Well, I like what I see. Will plugins not opening API fail at link time?

**Comment by matejkosik** [comment]
@ejgallego If a plugin tries to use something that we've not decided to make available, it will lead to compilation failure (of the plugin).

What is, in this branch, available to each plugin is:
- `API` module (provided by the `API/API.cma` library)
- `Grammar_API` module (provided by the `API/API.cma`)
- and any module provided by some other plugin.

**Comment by JasonGross** [comment]
Please export `cache_term_by_tactic_then` from `tactics.mli`; I plan to make use of it in fiat, and exported it specifically so that I could factor some of the code in fiat's plugin through it (because I need to get a handle on the generated name).

**Comment by matejkosik** [comment]
@maximedenes I see Travis results. Two tests failed. Both related to warnings (I think). What should I do? Remove `[@@ocaml.deprecated]`?

**Comment by maximedenes** [comment]
Ah, I don't know, good question. Will need to think about it. But there is also the fix asked by Jason.

**Comment by matejkosik** [comment]
@maximedenes Ok. Added. What about the warnings?

**Comment by SkySkimmer** [comment]
Making deprecation warnings into errors means we can't deprecate something in one PR and remove its internal uses in another PR, they have to be removed in the same travis round. This might not be the right choice.
OTOH this PR is producing thousands of deprecation warnings all over the plugins, plus the ones from API.mli, so I think it's overeager in deprecating
https://github.com/coq/coq/pull/652/files#diff-98ab24345b65563bb8de6c2d2e2dfa80R905 deprecated "unsafe" is a bit strange

NB: The unused value warning in API.mli is because of this pattern:
``` ocaml
val foo : t (* <- unused value *)
val foo : t
```
The other warnings (unused opens, unused types in API.ml, warning 40 which is because of insufficient opens) should be simple to fix.

**Comment by matejkosik** [comment]
On 05/20/2017 11:50 AM, Gaetan Gilbert wrote:
> Making deprecation warnings into errors means we can't deprecate something in one PR and remove its internal uses in another PR, they have to be removed in the same travis round.

Sorry, I do not understand what you mean above...

I read it as [@@ocaml.deprecated "..."] tags cannot be added by PRs. Is that what you mean?

What do you mean by "internal uses"?

What's "travis round"?

**Comment by SkySkimmer** [comment]
If you deprecated some function (with `[@@ocaml.deprecated]`) and that function is used somewhere in Coq (as opposed to some 3rd party plugin like ssreflect) then the travis test will trigger the warning and fail. So if you deprecate something you have to remove all uses of it before travis will accept the PR.

I should have typed "same PR" instead of "same travis round".

**Comment by ejgallego** [comment]
@SkySkimmer that strategy sounds fine to me in principle, it'd be good if deprecation patches were accompanied of adapting the code base.

**Comment by matejkosik** [comment]
@SkySkimmer Ok.

**Comment by SkySkimmer** [comment]
I'm still getting a number of warnings like 33 unused open and
>Warning 40: Pos was selected from type API.Sorts.contents.
It is not visible in the current scope, and will not 
be selected if the type becomes unknown.

**Comment by matejkosik** [comment]
I've looked what Travis is doing and tried to check consistently the same stuff locally. Now I am curious whether there is anything I still missed.

**Comment by ejgallego** [comment]
Indeed @matejkosik , are you using `-warn-error` locally? That should help you fix all the warnings.

**Comment by matejkosik** [comment]
@ejgallego Yes (with Ocaml `4.02.3` and with `4.04.1`).

**Comment by ejgallego** [comment]
@matejkosik if so, how did it happen that Travis failed?

**Comment by matejkosik** [comment]
@ejgallego I suppose I had a "dirty" state locally. If I did `make clean && make`, I would have seen that the there are some remaining warnings.

**Comment by matejkosik** [comment]
Ok, so warnings are now taken care of.

Let me know if you think there are other things that need to be done...
(and must be address within this pull-request instead of separate subsequent ones).

**Comment by matejkosik** [comment]
@maximedenes Do you think that the `fix_needed` label can be removed in this case?

**Comment by maximedenes** [comment]
Summary of the WG discussion:
- @ejgallego asks we think about the connection with the `META.coq` file (but says it's not blocking)
- @ejgallego shouldn't the API be compartmentalized component by component? made less monolithic?
- @silene what about documentation?
- @maximedenes what about stability? not constraining for now, document changes and functions that are changed
- @gares @ejgallego make it opt-out

**Comment by matejkosik** [comment]
@ejgallego The best answer (for the moment) I can formulate is [here](https://github.com/matejkosik/coq/blob/trunk__API/API/NOTES.md#shouldnt-the-api-be-compartmentalized-component-by-component-made-less-monolithic).

File: API/NOTES.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-Shouldn't the API be compartmentalized component by component? Made less monolithic?
+Shouldn't the API be compartmentalized [component](https://github.com/coq/coq/blob/trunk/META.coq) by [component](https://github.com/coq/coq/blob/trunk/META.coq)?
```
Comment (Zimmi48) [inline API/NOTES.md]: This is a completely trivial remark but why don't you do

    [component by component](https://github.com/coq/coq/blob/trunk/META.coq)

instead. With what you did instead, people will see two links and interested people may very well click on both, only to discover that it was twice the same link!

Comment (matejkosik) [inline API/NOTES.md]: Indeed. I've updated the file. Thanks!

**Comment by matejkosik** [comment]
On 05/23/2017 03:26 PM, Maxime Dénès wrote:
>   * @silene <https://github.com/silene> what about documentation?

Adding documentation is part of the plan but this pull-request does not have ambitions to do anything about it.

(The only realistic way to proceed seemed to be to make many small steps (each of which can be discussed separately). This is, I think, the first, smallest step in the right direction.

**Comment by matejkosik** [comment]
On 05/23/2017 03:26 PM, Maxime Dénès wrote:
> Summary of the WG discussion:
>
>   * @maximedenes <https://github.com/maximedenes> what about stability? not constraining for now, document changes and functions that are changed

To answer the above question, we should probably answer some other questions:
- How many people understand the current "API"? Who are they?
  (My optimistic estimate is that there are maybe 10--20 of them)
  - The question of stability is relevant to this set and this set of people only.
    It makes sense to talk to them and ask them what they think.
    We need to hear their opinions.

My humble efforts are trying to address the needs of the people who
- might want to understand Coq's API
- but are currently failing.

**Comment by matejkosik** [comment]
On 05/23/2017 03:26 PM, Maxime Dénès wrote:
> Summary of the WG discussion:
> 
>   * @gares <https://github.com/gares> @ejgallego <https://github.com/ejgallego> make it opt-out

If all goes well, i.e.
- we manage to define a meaningful API
- we, ourselves, will actually use it for our own plugins
then I think this policy (opt-out) probably will make sense.

**Comment by maximedenes** [comment]
It seems there are two commits "rebasing on trunk". How is that possible? Rebasing doesn't generate new commits.

**Comment by matejkosik** [comment]
It may be related to rebasing & conflict resolution. But you are right that the commits probably could be squashed. Should I do it?

**Comment by maximedenes** [comment]
I don't see what you mean. Were you rebasing or not?

**Comment by matejkosik** [comment]
I was rebasing and resolving conflicts (twice).

**Comment by maximedenes** [comment]
But anyway, I guess it would be ok to squash all commits in one, here.

**Comment by matejkosik** [comment]
Ok. I'll do that.

**Comment by maximedenes** [comment]
> I was rebasing and resolving conflicts (twice).

Then, where do the "rebasing on trunk" commits come from?

**Comment by maximedenes** [comment]
The last missing bit of this PR is to enforce the API abstraction via `coq_makefile` and provide an opt-out flag. Do you plan to do that in a separate PR or to add a commit to this one?

**Comment by matejkosik** [comment]
If possible, I'd like to do that in a separate pull-request because:
- our own plugins are not compiled via `coq_makefile`
- it specifically wanted to avoid mixing separable things together (and provide this functionality in a separate PR that can be discussed separately to avoid blocking this one).

**Comment by matejkosik** [comment]
Rebase complete. It triggered Travis which should give the exact same results as before (because I just squashed all commits together).

**Comment by maximedenes** [comment]
Ok, looks fine. Let's run the tests one last time and merge this PR. Thanks a lot!

**Comment by matejkosik** [comment]
Ok.

**Comment by matejkosik** [comment]
Traaavis! Traaavis! You can do it! Come on!

**Comment by Zimmi48** [comment]
@matejkosik Your invocation worked. The Travis build started just seconds ago.

**Comment by matejkosik** [comment]
Why is Travis checking commit `581939f6` instead of commit `0328bbb`?

**Comment by SkySkimmer** [comment]
Because it checks the result of merging with trunk.

**Comment by matejkosik** [comment]
Ok. Let me check ...

**Comment by Zimmi48** [comment]
I'm afraid @matejkosik that the build is failing due to bad interaction with #678, although the automatic merge was successful.

**Comment by matejkosik** [comment]
Travis checks for this commit [look good](https://travis-ci.org/matejkosik/coq/builds/237226256).

**Comment by matejkosik** [comment]
The first commit belongs to #700 on which I had to rebase this branch.

**Comment by matejkosik** [comment]
**TODO** : rebase this on trunk once #561 and #623 are merged.

**Comment by matejkosik** [comment]
rebase

**Comment by ejgallego** [comment]
> check why is it, currently, necessary to do open Stdarg in *.ml4 files? Can't macros generate fully-qualified identifiers?

This was explained by @ppedrot , IIRC indeed he considers such macros a bit fragile due to the bound to the module hierarchy, and I think he's right.

**Comment by matejkosik** [comment]
[This code](https://github.com/coq/coq/blob/a2a98a4015311af83edcf8fc87aa30a5318bead8/grammar/argextend.mlp#L22) is one of those that generates non-fully-qualified Ocaml code.

However, straightforward change to:
```
let mk_extraarg loc s = <:expr< API.Stdarg.$lid:"wit_"^s$ >>
```
is not possible, because even though sometimes the generated identifiers are really in `Stdarg`, othertimes they are provided by other modules.

I don't fully understand the details but it does not feel right.

**Comment by ppedrot** [comment]
@matejkosik My argument for not adding `Stdarg` qualification automatically is precisely what you describe: it should be easy for the plugin writer to understand where the argument name is taken. In practice, it just looks for the one whose name is statically bound by the OCaml code, which means that we don't have to special case any generic argument.

**Comment by matejkosik** [comment]
> it should be easy for the plugin writer to understand where the argument name is taken.

The current situation is that in order to use the macros, the plugin writer needs to read their implementation (in order to understand why they the generated code does not compile).

Why do you think this is better than defining macros which would expand to Ocaml code without unbound variables? @ejgallego @maximedenes  Do you understand Piere-Marie's point? It is a petty detail but one of those I never understood :-/

**Comment by ejgallego** [comment]
I think that indeed this is a subtle point, however I think that for 8.7 we should stick to the current scheme. But feel free to discuss / CEP the current macro / genarg stuff in the next WG, I'm sure can improve things.

**Comment by matejkosik** [comment]
Ok.

**Comment by matejkosik** [comment]
@maximedenes If `trunk` now [fails on Travis](https://travis-ci.org/coq/coq/builds/240011079) and this PR is rebased on `trunk`, it will also fail. I am wondering what should I do now ...

**Comment by maximedenes** [comment]
@matejkosik Trunk fails because we are waiting for some upstream merges, no big deal.

**Comment by matejkosik** [comment]
Green lights here: https://travis-ci.org/matejkosik/coq/builds/240247741

**Comment by matejkosik** [comment]
@maximedenes I think what you see is a bug in github, I think.

In this pull-request, the set of commits is the right one, but the order is not the same as in the [original branch](https://github.com/matejkosik/coq/commits/trunk__API) where I have moved "adding overlays" at the end.

If you merged this as is, I wouldn't object. If you want me do do something more (like dropping the commit related to overlays), let me know.

I think the results [produced by Travis](https://travis-ci.org/matejkosik/coq/builds/240356872) are relevant for this branch because those commits which are swapped are unrelated.

**Comment by Zimmi48** [comment]
@matejkosik I have seen this bug (wrong order of commits) a lot these days.

**Comment by matejkosik** [comment]
@Zimmi48 That's interesting.

**Comment by matejkosik** [comment]
@maximedenes VST was [broken upstream](https://travis-ci.org/JasonGross/coq/jobs/240521141#L4176) at the time when Travis tried to compile it. Is there anything else I have to do here or can you already merge this PR?

**Comment by matejkosik** [comment]
@maximedenes I tried right now to compile VST with vanilla Coq `trunk` I am getting the same error:
```
COQC veric/jstep.v
COQC veric/semax_ext_oracle.v
COQC sepcomp/event_semantics.v
COQC veric/mem_lessdef.v
COQC veric/aging_lemmas.v
COQC floyd/coqlib3.v
COQC floyd/base.v
File "./floyd/base.v", line 5, characters 15-32:
Error: Unable to locate library floyd.jmeq_lemmas.

make[1]: *** [floyd/base.vo] Error 1
make[1]: Leaving directory `/nas/home1/m/mkosik/git/coq/trunk0/_build_ci/VST'
make: *** [ci-vst] Error 1
```
The error is different but also the commit of VST is different. In any case, VST still remains broken.

**Comment by tlringer** [comment]
I know this is an old commit, but I can't find the place where the major API discussion happened. Is there a reason the following aren't exposed:

nb_rel
whd_betadeltaiota
pop_rel_context
string_of_kn
pp_with

These are all functions that my plugin calls. Are these situations where I should be calling something else, or are these safe? (Some of the names may be outdated, since my plugin is from 8.5.)

**Comment by JasonGross** [comment]
I believe discussion on the API was scattered (this PR, perhaps other PRs, maybe a CEP, definitely gitter, maybe the video logs of the CoqWG), and there was no particular coherence, as far as I could tell, to what went in to the API and what didn't.  Submit a PR adding those to the API, and I expect it'll get merged pretty quickly.  In the meantime, I believe `coq_makefile` accepts `-bypass-API`.  It might also be worth it to add your development to Coq's CI, so silent breaks like this don't happen again.

**Comment by SkySkimmer** [comment]
The content of the API is largely guided by what the plugins we were looking at use, essentially meaning the plugins distributed with Coq plus whatever the developments tested in our travis use.

As @JasonGross said you can compile with `-bypass-API` to avoid API problems while dealing with other changes. Alternatively you can target 8.7 (currently beta for a couple more weeks) which doesn't have the API system.

As to the specific values you mentioned:

`string_of_kn` is replaced by `Names.KerName.to_string`
`pp_with` is from lib/pp, all modules in lib/ are not hidden by the API so are fully available (I'm not familiar with the doc for API.mli, maybe this could be made clearer?)
`whd_betadeltaiota` is now called `whd_all` https://github.com/coq/coq/blob/master/dev/doc/changes.md#reduction-functions (note that there is one in reductionops and one in reduction, depending on whether you're working with [econstr](https://github.com/coq/coq/blob/master/dev/doc/econstr.md) or kernel constr.)

`nb_rel` and `pop_rel_context` can be added as far as I can tell, just confirm you want them or make the PR yourself.

**Comment by ejgallego** [comment]
@tlringer most of the functions you mention are not affected by "the API" commit itself, but as @SkySkimmer has pointed out the functions have been removed or renamed. 8.5 depended on an old OCaml version that didn't support proper deprecation notices; the situation should be better now.

As for `pp_with` the function is still there however the signature may have changed from 8.5.

**Comment by ejgallego** [comment]
Targetting 8.7 is a very good call; adding your PR to the CI is another one.

**Comment by tlringer** [comment]
OK, thanks. My concern right now is mostly with safety. I'd love nb_rel and pop_rel_context. If you add them now that's great, if not I can submit a PR after my next deadline. Thanks for the prompt feedback.

**Comment by ejgallego** [comment]
> I can submit a PR after my deadline next week

I'd say don't bother with the PR yet; as noted before, 8.7 doesn't use the API and it could be well the case that for 8.8 we end up with a different solution.

**Pull request merged**